actioncable/app/javascript/action_cable/action_cable.js

```diff
@@ -0,0 +1,43 @@
+import Connection from "./connection"
+import ConnectionMonitor from "./connection_monitor"
+import Consumer from "./consumer"
+import INTERNAL from "./internal"
+import { log, logger } from "./logger"
+import Subscription from "./subscription"
+import Subscriptions from "./subscriptions"
+
+export function getConfig(name) {
+  const element = document.head.querySelector(`meta[name='action-cable-${name}']`)
+  return (element ? element.getAttribute("content") : undefined)
+}
+
+export function createWebSocketURL(url) {
+  if (url && !/^wss?:/i.test(url)) {
+    const a = document.createElement("a")
+    a.href = url
+    // Fix populating Location properties in IE. Otherwise, protocol will be blank.
+    a.href = a.href
+    a.protocol = a.protocol.replace("http", "ws")
+    return a.href
+  } else {
+    return url
+  }
+}
+
+export function createConsumer(url) {
+  if (url == null) {
+    const urlConfig = getConfig("url")
+    url = (urlConfig ? urlConfig : INTERNAL.default_mount_path)
+  }
+  return new Consumer(createWebSocketURL(url))
+}
+
+export function startDebugging() {
+  logger.enabled = true
+}
+
+export function stopDebugging() {
+  logger.enabled = false
+}
+
+export { Connection, ConnectionMonitor, Consumer, INTERNAL, log, Subscription, Subscriptions }
```

Comment:
Very nice! I like this approach to exporting the `ActionCable` interface much more than https://github.com/rails/rails/blob/ae122c32c11336734dc86d7bbdfdea5bfda2b6d6/actioncable/app/javascript/action_cable/index.js#L54-L61

The one thing that's missing is access to the `WebSocket` adapter here. e.g. being able to set `ActionCable.WebSocket = …` in an application.

Reply:
The syntax you’re referencing will still work if the application does:
```js
import ActionCable from "actioncable"
```
or uses the UMD export of the compiled bundle.

This behavior is verified by one of the existing tests: https://github.com/rails/rails/blob/405aef3dede8938c8eb9789f6d6722864faddb09/actioncable/test/javascript/src/unit/action_cable_test.coffee#L10-L12

If an application is using fine grained imports, they can opt to instead set the adapter via:
```js
import adapters from "actioncable/adapters"

adapters.WebSocket = MyAdapter
```

Reply:
Further conversation moved to https://github.com/rails/rails/pull/34177#discussion_r229398342

Reply:
Moving the discussion back here.

> > I support keeping `index.js` and `cable.js` separate because if we combined them, we'd lose the ability for applications to tree-shake unused functions from ActionCable. I don't see why we should sacrifice that in order to have one fewer source file.
> 
> Can you share an example of that? How would you cherry-pick your imports for optimal tree-shaking with those separate files

I'm not sure if @sveinn has his own example, but the example repository I created demonstrates this: https://github.com/rmacklin/actioncable-es2015-build-example

Here's the way to cherry-pick the imports for optimal tree-shaking:
https://github.com/rmacklin/actioncable-es2015-build-example/commit/d494d6bf160042ca218bf3bd7aae6c1929492e0b#diff-4a23ac1286faa3273c8cdc9b4bb5078d

The difference in the unminified generated code is:
1. The assignment of `getConfig`, `createWebSocketURL`, `createConsumer`, `startDebugging`, `stopDebugging`, `Connection`, `ConnectionMonitor`, `Consumer`, `INTERNAL`, `log`, `Subscription`, `Subscriptions` onto `action_cable_namespaceObject` gets removed:
   https://github.com/rmacklin/actioncable-es2015-build-example/commit/831b2b90e6b0c072eabe1116b032b52d5735e416#diff-d2679f9a86538b3db8ad2a97579f226aL8
2. The delegated getters and setters for `ActionCable.logger` and `ActionCable.WebSocket` are not used in the application, so they get removed:
   https://github.com/rmacklin/actioncable-es2015-build-example/commit/831b2b90e6b0c072eabe1116b032b52d5735e416#diff-d2679f9a86538b3db8ad2a97579f226aL611

   And there are additional differences in when the optimized code is minified:

3. `createConsumer` is inlined into an anonymous function in the only place that it is used:
   https://github.com/rmacklin/actioncable-es2015-build-example/commit/57c4abe061ee6aad55e44d14592f59887f436566#diff-363285dc602b7237f78d880e26c875a0R348
4. `stopDebugging` and `startDebugging` are no longer referenced as a result of (1), so they get removed:
   https://github.com/rmacklin/actioncable-es2015-build-example/commit/57c4abe061ee6aad55e44d14592f59887f436566#diff-363285dc602b7237f78d880e26c875a0L29

Reply:
> and what would the difference be if they were combined?

If the files are combined you cannot preserve the backwards compatible API and also allow `createConsumer` to be imported individually. But I personally have no problem with changing the public API, for [these reasons](https://github.com/rails/rails/pull/34177#discussion_r229741614):

> However, if we don't want to have the two separate source files, there is an alternative that I would be fine with: We could also export `adapters` from `action_cable.js` and move `action_cable.js` to `index.js`. Then all functionality would be available from the individual exports of the new `index.js` (former `action_cable.js`), but we would be slightly changing the public API. I don't think that's a huge cause for concern because:
> 
> * We can share the contents of what is currently `index.js` in the upgrade guide as a way for users to patch the old API into their applications when they upgrade, until they migrate to the new API.
> * We could go further and ship a codemod that they can run to automatically translate their application to the new API. (If you're not familiar, this is something [react has been doing](https://github.com/reactjs/react-codemod) to make API upgrades easy for users.) I'd be happy to look into writing the codemod.

Reply:
@javan if I'm interpreting [your last comment](https://github.com/rails/rails/pull/34177#discussion_r229832626) correctly, it sounds like you're okay with modifying the public API by only offering individual exports instead of exporting an `ActionCable` object containing everything. Is that right?

If so I'll refactor this PR to remove `action_cable.js` and only offer individual exports from `index.js`.

Reply:
I've opened a PR with the alternate approach so we can easily compare them: https://github.com/rails/rails/pull/34370. I'm happy to close either PR in favor of the other, depending on which approach is preferred.